[scripts] Add trainer option --trainer.optimization.num-jobs-step; python3 compat fixes

kkm000 opened the pull request:
1. Add option `--trainer.optimization.num-jobs-step [N]` (default:1) to
specify the multiple of the number of jobs when the computed requested
number exceeds this number. This is intended for training on a limited
number of GPUs, to keep them all efficiently loaded. The rounding is
such that the total number of spawned jobs is close to the current
behavior, when this increment equals 1.

For example, when this option set 4 and the initial-jobs=2, the progress
in the number of jobs will be {2, 3, 4, 8, 12...} etc. Of course, each
number lingers for a proportionally longer number of iterations, for
example, for consecutive iterations, {...4, 4, 4, 4, 4, 8, 8,...} where
the default setting of 1 would be {...4, 5, 6, 7, 8, 8, 9, ....} (the
test in common.py has an example). The sum of these numbers with the
step N differs from the default case N_0=1 by no more than
int(N/2 + 0.5).

The current number of jobs is now added to the iteration log printout.

2. Fix the validate_minibatch_size_str to work in poth python 2 and 3,
and add unit tests for it. It failed because python3 does not implicitly
convert str to int in a comparison.

3. Add a few missing .format() to supply arguments to format strings in
error messages.

Close #3186

kkm000 commented:
Sheesh... Looks like I'm out of date on python3: https://www.python.org/dev/peps/pep-3113/ -- and I was kinda used to this feature. Even liked it. But I've never been a string-going pythonista tho. Will fix. WIPing this.

1965654167 commented:
你好

kkm000 commented:
嘿，智能机器人先生! But you are already straining the limits of my Chinese, sorry...

danpovey commented:
We are looser on the num-chars-per-line for python, around 100 is OK but
it's not super strict.

@vimalmanohar can you please look over this PR with a view mainly to
ensuring that it wouldn't break any existing code?  Just a few minutes of
your time is all that's needed, I have a lot to catch up on as I flew to
Redmond yesterday.


On Thu, Apr 4, 2019 at 5:54 AM kkm (aka Kirill Katsnelson) <
notifications@github.com> wrote:

> 嘿，智能机器人先生! But you are already straining the limits of my Chinese, sorry...
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/3205#issuecomment-479885056>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVuwii4LC_S_3eM-WEFDlLCyFtI-hOks5vdfYhgaJpZM4cbpIE>
> .
>

danpovey commented:
@huangruizhe can you please try to run the current mini-librispeech chain TDNN training example with this version of the scripts?  This is just to make sure that it doesn't break anything in the existing setup.

huangruizhe commented:
Sure.

kkm000 commented:
@huangruizhe, @danpovey: I actually did, although starting with existing egs. I am too careful. The line I posted was from the log; I can just attach the whole thing and save you the trouble, but it was totally ok. Also see below:

@vimalmanohar, I used a python2 feature which they killed in python3. So I'll have to fix it. And I am all for upgrading the training scripts to python3, as we are requiring it anyway, mainly because of the `f'...'` interpolating strings. The formatting part is probably one of the lest readable parts of the current scripts.

I only need to write a maintenance script to automatically apply the diff to other train* files. They are 90% copypasta, including some minor bugs I noticed but only after I replicated this patch by hand...

danpovey commented:
So Kirill, am I to understand that this is not quite ready for merge due to
that python3 issue?

On Thu, Apr 4, 2019 at 3:55 PM kkm (aka Kirill Katsnelson) <
notifications@github.com> wrote:

> @huangruizhe <https://github.com/huangruizhe>, @danpovey
> <https://github.com/danpovey>: I actually did, although starting with
> existing egs. I am too careful. The line I posted was from the log; I can
> just attach the whole thing and save you the trouble, but it was totally
> ok. Also see below:
>
> @vimalmanohar <https://github.com/vimalmanohar>, I used a python2 feature
> which they killed in python3. So I'll have to fix it. And I am all for
> upgrading the training scripts to python3, as we are requiring it anyway,
> mainly because of the f'...' interpolating strings. The formatting part
> is probably one of the lest readable parts of the current scripts.
>
> I only need to write a maintenance script to automatically apply the diff
> to other train* files. They are 90% copypasta, including some minor bugs I
> noticed but only after I replicated this patch by hand...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/3205#issuecomment-480092090>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu9-r-Zki6WAcJIRtInAVwmK1RW76ks5vdoLkgaJpZM4cbpIE>
> .
>

kkm000 commented:
@danpovey -- no, not yet, please hold. I had to mark it WIP. I tried unit tests with python3, and the compile blew up on a syntax error. Then I found why. I rate Python the most bleak and dreary language in the current wide use (after BASIC and Fortran, neé FORTRAN, but they are more an exotics now), and not they took *even that* from me!

I'll post a log after I change the parameters, you probably do not need to rerun; I will do anyway. Whether this setting affects the end result, I do not know. In my experience, any change is likely to cause a small variation. I'm still trying to figure out the empiric relation of the number of jobs to xent_regularize (fewer parallel jobs require a higher xent_regularize; looks like the averaging is a regularization by itself).

Log excerpt (1 epoch, LR makes no sense, ran just co confirm it technically runs to the end)
```
 'num_epochs': 1.0,
 'num_jobs_final': 10,
 'num_jobs_initial': 2,
 'num_jobs_step': 2,
 . . .
2019-04-03 03:33:26,561 [steps/nnet3/chain/train.py:411 - train - INFO ] Copying the properties from exp/gsm1/chain/train_clean_100_hires/res100a-resmart01/egs to exp/gsm1/chain/train_clean_100_hires/res100a-smart01-temp
2019-04-03 03:33:26,562 [steps/nnet3/chain/train.py:468 - train - INFO ] Training will run for 1.0 epochs = 24 iterations
2019-04-03 03:33:26,562 [steps/nnet3/chain/train.py:512 - train - INFO ] Iter:  3/23   Jobs: 4   Epoch: 0.04/1.0 (4.1% complete)    lr: 0.001029
2019-04-03 03:36:15,391 [steps/nnet3/chain/train.py:512 - train - INFO ] Iter:  4/23   Jobs: 4   Epoch: 0.07/1.0 (6.8% complete)    lr: 0.000929
2019-04-03 03:39:09,887 [steps/nnet3/chain/train.py:512 - train - INFO ] Iter:  5/23   Jobs: 4   Epoch: 0.10/1.0 (9.5% complete)    lr: 0.000839
2019-04-03 03:41:59,494 [steps/nnet3/chain/train.py:512 - train - INFO ] Iter:  6/23   Jobs: 4   Epoch: 0.12/1.0 (12.2% complete)    lr: 0.000757
2019-04-03 03:44:48,877 [steps/nnet3/chain/train.py:512 - train - INFO ] Iter:  7/23   Jobs: 4   Epoch: 0.15/1.0 (15.0% complete)    lr: 0.000684
2019-04-03 03:47:37,805 [steps/nnet3/chain/train.py:512 - train - INFO ] Iter:  8/23   Jobs: 4   Epoch: 0.18/1.0 (17.7% complete)    lr: 0.000617
2019-04-03 03:50:26,892 [steps/nnet3/chain/train.py:512 - train - INFO ] Iter:  9/23   Jobs: 6   Epoch: 0.20/1.0 (20.4% complete)    lr: 0.000836
 . . . nvidia not in exclusive mode kaboom, restart ...
 'stage': 9,
 . . . 
2019-04-03 03:54:00,298 [steps/nnet3/chain/train.py:410 - train - INFO ] Copying the properties from exp/gsm1/chain/train_clean_100_hires/res100a-resmart01/egs to exp/gsm1/chain/train_clean_100_hires/res100a-smart01-temp
2019-04-03 03:54:00,298 [steps/nnet3/chain/train.py:467 - train - INFO ] Training will run for 1.0 epochs = 24 iterations
2019-04-03 03:54:00,298 [steps/nnet3/chain/train.py:511 - train - INFO ] Iter: 9/23   Jobs: 6   Epoch: 0.20/1.0 (20.4% complete)   lr: 0.000836
2019-04-03 03:58:26,293 [steps/nnet3/chain/train.py:511 - train - INFO ] Iter: 10/23   Jobs: 6   Epoch: 0.24/1.0 (24.5% complete)   lr: 0.000717
2019-04-03 04:02:52,917 [steps/nnet3/chain/train.py:511 - train - INFO ] Iter: 11/23   Jobs: 6   Epoch: 0.29/1.0 (28.6% complete)   lr: 0.000615
2019-04-03 04:07:20,886 [steps/nnet3/chain/train.py:511 - train - INFO ] Iter: 12/23   Jobs: 6   Epoch: 0.33/1.0 (32.7% complete)   lr: 0.000528
2019-04-03 04:11:48,058 [steps/nnet3/chain/train.py:511 - train - INFO ] Iter: 13/23   Jobs: 6   Epoch: 0.37/1.0 (36.7% complete)   lr: 0.000453
2019-04-03 04:16:08,760 [steps/nnet3/chain/train.py:511 - train - INFO ] Iter: 14/23   Jobs: 6   Epoch: 0.41/1.0 (40.8% complete)   lr: 0.000388
2019-04-03 04:20:31,957 [steps/nnet3/chain/train.py:511 - train - INFO ] Iter: 15/23   Jobs: 8   Epoch: 0.45/1.0 (44.9% complete)   lr: 0.000444
2019-04-03 04:26:18,238 [steps/nnet3/chain/train.py:511 - train - INFO ] Iter: 16/23   Jobs: 8   Epoch: 0.50/1.0 (50.3% complete)   lr: 0.000362
2019-04-03 04:32:09,493 [steps/nnet3/chain/train.py:511 - train - INFO ] Iter: 17/23   Jobs: 8   Epoch: 0.56/1.0 (55.8% complete)   lr: 0.000295
2019-04-03 04:38:06,404 [steps/nnet3/chain/train.py:511 - train - INFO ] Iter: 18/23   Jobs: 8   Epoch: 0.61/1.0 (61.2% complete)   lr: 0.000240
2019-04-03 04:43:58,228 [steps/nnet3/chain/train.py:511 - train - INFO ] Iter: 19/23   Jobs: 8   Epoch: 0.67/1.0 (66.7% complete)   lr: 0.000196
2019-04-03 04:49:49,639 [steps/nnet3/chain/train.py:511 - train - INFO ] Iter: 20/23   Jobs: 8   Epoch: 0.72/1.0 (72.1% complete)   lr: 0.000160
2019-04-03 04:56:08,594 [steps/nnet3/chain/train.py:511 - train - INFO ] Iter: 21/23   Jobs: 10   Epoch: 0.78/1.0 (77.6% complete)   lr: 0.000163
2019-04-03 05:03:20,491 [steps/nnet3/chain/train.py:511 - train - INFO ] Iter: 22/23   Jobs: 10   Epoch: 0.84/1.0 (84.4% complete)   lr: 0.000126
2019-04-03 05:10:36,747 [steps/nnet3/chain/train.py:511 - train - INFO ] Iter: 23/23   Jobs: 10   Epoch: 0.91/1.0 (91.2% complete)   lr: 0.000070
2019-04-03 05:17:49,610 [steps/nnet3/chain/train.py:567 - train - INFO ] Doing final combination to produce final.mdl
2019-04-03 05:17:49,610 [steps/libs/nnet3/train/chain_objf/acoustic_model.py:571 - combine_models - INFO ] Combining set([13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24]) models.
2019-04-03 05:18:09,571 [steps/nnet3/chain/train.py:596 - train - INFO ] Cleaning up the experiment directory exp/gsm1/chain/train_clean_100_hires/res100a-smart01-temp
exp/gsm1/chain/train_clean_100_hires/res100a-smart01-temp: num-iters=24 nj=2..10 num-params=14.2M dim=40+100->3032 combine=-0.056->-0.056 (over 1) xent:train/v

kkm000 commented:
I'll do my best to send in the fix today.

kkm000 commented:
Ah, and `halve_minibatch_size_str` fails under py3 too

kkm000 commented:
So I'll fix it too. No implicit str -> int any more.
```
$ python3 -c $'i="128"\nif i > 0:\n print(i)'
Traceback (most recent call last):
  File "<string>", line 2, in <module>
TypeError: '>' not supported between instances of 'str' and 'int'
$ python2 -c $'i="128"\nif i > 0:\n print(i)'
128
```

kkm000 commented:
@vimalmanohar, PTAL. I fixed the py3 compatibility, and also found and fixed another issue in `validate_minibatch_size_str` which was failing under py3, because it does not allow str->int in a comparison, `if '123' <= 0:` does not work. Also added unit tests for this function, with both success and failure cases, tried to be pretty exhaustive. All unit tests pass under both py2 and py3. Incidentally, there were also a few strings with a `"{0}"` without a `.format()` after them, added these too.

I quickly ran the chain/train.py from start to end with a very few iterations successfully (LR makes no sense, I swapped lr-final and lr-inital by mistake, so don't be alarmed it grows, totally expected).

```
 . . . abridged . . .
 'final_effective_lrate': 0.007,
 'initial_effective_lrate': 0.0003,
 'num_epochs': 0.28,
 'num_jobs_final': 10,
 'num_jobs_initial': 2,
 'num_jobs_step': 2,
. . . . .
2019-04-04 18:34:16,495 [steps/nnet3/chain/train.py:467 - train - INFO ] Training will run for 0.28 epochs = 6 iterations
2019-04-04 18:34:16,495 [steps/nnet3/chain/train.py:511 - train - INFO ] Iter: 0/5   Jobs: 2   Epoch: 0.00/0.3 (0.0% complete)   lr: 0.000600
2019-04-04 18:35:52,150 [steps/nnet3/chain/train.py:511 - train - INFO ] Iter: 1/5   Jobs: 4   Epoch: 0.01/0.3 (4.9% complete)   lr: 0.001399
2019-04-04 18:38:42,150 [steps/nnet3/chain/train.py:511 - train - INFO ] Iter: 2/5   Jobs: 4   Epoch: 0.04/0.3 (14.6% complete)   lr: 0.001903
2019-04-04 18:41:31,759 [steps/nnet3/chain/train.py:511 - train - INFO ] Iter: 3/5   Jobs: 6   Epoch: 0.07/0.3 (24.4% complete)   lr: 0.003881
2019-04-04 18:45:53,472 [steps/nnet3/chain/train.py:511 - train - INFO ] Iter: 4/5   Jobs: 8   Epoch: 0.11/0.3 (39.0% complete)   lr: 0.008205
2019-04-04 18:51:39,691 [steps/nnet3/chain/train.py:511 - train - INFO ] Iter: 5/5   Jobs: 8   Epoch: 0.16/0.3 (58.5% complete)   lr: 0.056000
2019-04-04 18:57:31,720 [steps/nnet3/chain/train.py:567 - train - INFO ] Doing final combination to produce final.mdl
2019-04-04 18:57:31,720 [steps/libs/nnet3/train/chain_objf/acoustic_model.py:571 - combine_models - INFO ] Combining set([4, 5, 6]) models.
2019-04-04 18:57:39,943 [steps/nnet3/chain/train.py:596 - train - INFO ] Cleaning up the experiment directory exp/gsm1/chain/train_clean_100_hires/res100a-smart01-temp
exp/gsm1/chain/train_clean_100_hires/res100a-smart01-temp: num-iters=24 nj=2..10 num-params=14.2M dim=40+100->3032 combine=-0.095->-0.095 (over 1) xent:train/valid[15,23]=(-1.27,-1.15/-1.31,-1.23) logprob:train/valid[15,23]=(-0.070,-0.058/-0.077,-0.071)
```

kkm000 commented:
And the default behavior does not change, with the default num-jobs-step of 1 the number of jobs is computes exactly as it was. This is for training on a machine with N GPUs, so after the current num jobs reaches the number of GPU, more iterations are run with the same number of jobs, and then it jumps right to 2×step and so on. So that all of the GPUs are churning :)

danpovey commented:
Great.  Let me know when you think it's ready, I may have someone at
Hopkins run it here just to double check.

On Thu, Apr 4, 2019 at 10:56 PM kkm (aka Kirill Katsnelson) <
notifications@github.com> wrote:

> And the default behavior does not change, with the default num-jobs-step
> of 1 the number of jobs is computes exactly as it was. This is for training
> on a machine with N GPUs, so after the current num jobs reaches the number
> of GPU, more iterations are run with the same number of jobs, and then it
> jumps right to 2×step and so on. So that all of the GPUs are churning :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/3205#issuecomment-480157341>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVuy4VDOKlpjLTAJQXyL2Z0KKMkn9Mks5vduWpgaJpZM4cbpIE>
> .
>

kkm000 commented:
It is. If you test, I did chain/train.py only, so you might want to run e2e/chain/train.py or others, just to cover  more.

huangruizhe commented:
I am running the mini_librispeech recipe as a double check. It seems it covers `steps/nnet3/chain/train.py` only, should I run any recipe to test `steps/nnet3/chain/e2e/train_e2e.py` also? @danpovey

danpovey commented:
If you have time to run the e2e code too it would be great; I'm not in a
huge hurry to merge this in the next few days.


On Fri, Apr 5, 2019 at 12:35 PM huangruizhe <notifications@github.com>
wrote:

> I am running the mini_librispeech recipe as a double check. It seems it
> covers steps/nnet3/chain/train.py only, should I run any recipe to test
> steps/nnet3/chain/e2e/train_e2e.py also? @danpovey
> <https://github.com/danpovey>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/3205#issuecomment-480397164>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu08EfQ4Hh9QgTfDEsR1lqtl81i2vks5vd6V1gaJpZM4cbpIE>
> .
>

kkm000 commented:
Up to you. The train*.py files are so similar the diff applied cleanly to all 6 without fuzz, only with offsets. If one is ok, all are.

huangruizhe commented:
Oh, I see...

kkm000 commented:
Ehm... Did it pass?

huangruizhe commented:
The `train.py` has passed. I am running the `wsj/local/e2e/run_end2end_phone.sh` for `train_e2e.py`, will know how it goes later.

huangruizhe commented:
I got an error like this, I am not sure whether it is related to the PR:
```
$ bash local/e2e/run_end2end_phone.sh >> log 2>&1
...
2019-04-13 08:49:36,180 [steps/nnet3/chain/e2e/train_e2e.py:460 - train - INFO ] Iter: 24/179   Jobs: 3   Epoch: 0.63/10.0 (6.3% complete)   lr: 0.001296
queue.pl: job failed with status 134, log is in exp/chain/e2e_tdnnf_1a/log/train.24.1.log
```
Looking into the train.24.1.log:
```
# Running on b05
# Started at Sat Apr 13 08:53:52 EDT 2019
# nnet3-chain-train --use-gpu=yes --apply-deriv-weights=False --l2-regularize=5e-05 --leaky-hmm-coefficient=0.1 --read-cache=exp/chain/e2e_tdnnf_1a/cache.24 --write-cache=exp/chain/e2e_tdnnf_1a/cache.25 --xent-regularize=0.0 --print-interval=10 --momentum=0.0 --max-param-change=2.0 --backstitch-training-scale=0.0 --backstitch-training-interval=1 --l2-regularize-factor=0.333333333333 --srand=24 "nnet3-am-copy --raw=true --learning-rate=0.00129645642911 --scale=1.0 exp/chain/e2e_tdnnf_1a/24.mdl - |nnet3-copy --edits='set-dropout-proportion name=* proportion=0.0' - - |" exp/chain/e2e_tdnnf_1a/den.fst "ark,bg:nnet3-chain-copy-egs                          --frame-shift=2                         ark:exp/chain/e2e_tdnnf_1a/egs/cegs.28.ark ark:- |                         nnet3-chain-shuffle-egs --buffer-size=5000                         --srand=24 ark:- ark:- | nnet3-chain-merge-egs                         --minibatch-size=150=128,64/300=64,32/600=32,16/1200=8 ark:- ark:- |" exp/chain/e2e_tdnnf_1a/25.1.raw
nnet3-chain-train --use-gpu=yes --apply-deriv-weights=False --l2-regularize=5e-05 --leaky-hmm-coefficient=0.1 --read-cache=exp/chain/e2e_tdnnf_1a/cache.24 --write-cache=exp/chain/e2e_tdnnf_1a/cache.25 --xent-regularize=0.0 --print-interval=10 --momentum=0.0 --max-param-change=2.0 --backstitch-training-scale=0.0 --backstitch-training-interval=1 --l2-regularize-factor=0.333333333333 --srand=24 "nnet3-am-copy --raw=true --learning-rate=0.00129645642911 --scale=1.0 exp/chain/e2e_tdnnf_1a/24.mdl - |nnet3-copy --edits='set-dropout-proportion name=* proportion=0.0' - - |" exp/chain/e2e_tdnnf_1a/den.fst 'ark,bg:nnet3-chain-copy-egs                          --frame-shift=2                         ark:exp/chain/e2e_tdnnf_1a/egs/cegs.28.ark ark:- |                         nnet3-chain-shuffle-egs --buffer-size=5000                         --srand=24 ark:- ark:- | nnet3-chain-merge-egs                         --minibatch-size=150=128,64/300=64,32/600=32,16/1200=8 ark:- ark:- |' exp/chain/e2e_tdnnf_1a/25.1.raw
*** Error in `nnet3-chain-train': double free or corruption (!prev): 0x000056282f6f3440 ***
======= Backtrace: =========
...
```
I will ask someone else whether this is random and what the problem is.

danpovey commented:
No, that's related to a bug in the current NVidia drivers when they run on
a particular type of older hardware.  We generally run using cmd="retry.pl
--num-retries 3 queue.pl" to avoid it killing the entire run.


On Sat, Apr 13, 2019 at 3:10 AM huangruizhe <notifications@github.com>
wrote:

> I got an error like this, I am not sure whether it is related to the PR:
>
> $ bash local/e2e/run_end2end_phone.sh >> log 2>&1
> ...
> 2019-04-13 08:49:36,180 [steps/nnet3/chain/e2e/train_e2e.py:460 - train - INFO ] Iter: 24/179   Jobs: 3   Epoch: 0.63/10.0 (6.3% complete)   lr: 0.001296queue.pl: job failed with status 134, log is in exp/chain/e2e_tdnnf_1a/log/train.24.1.log
>
> Looking into the train.24.1.log:
>
> # Running on b05
> # Started at Sat Apr 13 08:53:52 EDT 2019
> # nnet3-chain-train --use-gpu=yes --apply-deriv-weights=False --l2-regularize=5e-05 --leaky-hmm-coefficient=0.1 --read-cache=exp/chain/e2e_tdnnf_1a/cache.24 --write-cache=exp/chain/e2e_tdnnf_1a/cache.25 --xent-regularize=0.0 --print-interval=10 --momentum=0.0 --max-param-change=2.0 --backstitch-training-scale=0.0 --backstitch-training-interval=1 --l2-regularize-factor=0.333333333333 --srand=24 "nnet3-am-copy --raw=true --learning-rate=0.00129645642911 --scale=1.0 exp/chain/e2e_tdnnf_1a/24.mdl - |nnet3-copy --edits='set-dropout-proportion name=* proportion=0.0' - - |" exp/chain/e2e_tdnnf_1a/den.fst "ark,bg:nnet3-chain-copy-egs                          --frame-shift=2                         ark:exp/chain/e2e_tdnnf_1a/egs/cegs.28.ark ark:- |                         nnet3-chain-shuffle-egs --buffer-size=5000                         --srand=24 ark:- ark:- | nnet3-chain-merge-egs                         --minibatch-size=150=128,64/300=64,32/600=32,16/1200=8 ark:- ark:- |" exp/chain/e2e_tdnnf_1a/25.1.raw
> nnet3-chain-train --use-gpu=yes --apply-deriv-weights=False --l2-regularize=5e-05 --leaky-hmm-coefficient=0.1 --read-cache=exp/chain/e2e_tdnnf_1a/cache.24 --write-cache=exp/chain/e2e_tdnnf_1a/cache.25 --xent-regularize=0.0 --print-interval=10 --momentum=0.0 --max-param-change=2.0 --backstitch-training-scale=0.0 --backstitch-training-interval=1 --l2-regularize-factor=0.333333333333 --srand=24 "nnet3-am-copy --raw=true --learning-rate=0.00129645642911 --scale=1.0 exp/chain/e2e_tdnnf_1a/24.mdl - |nnet3-copy --edits='set-dropout-proportion name=* proportion=0.0' - - |" exp/chain/e2e_tdnnf_1a/den.fst 'ark,bg:nnet3-chain-copy-egs                          --frame-shift=2                         ark:exp/chain/e2e_tdnnf_1a/egs/cegs.28.ark ark:- |                         nnet3-chain-shuffle-egs --buffer-size=5000                         --srand=24 ark:- ark:- | nnet3-chain-merge-egs                         --minibatch-size=150=128,64/300=64,32/600=32,16/1200=8 ark:- ark:- |' exp/chain/e2e_tdnnf_1a/25.1.raw
> *** Error in `nnet3-chain-train': double free or corruption (!prev): 0x000056282f6f3440 ***
> ======= Backtrace: =========
> ...
>
> I will ask someone else whether this is random and what the problem is.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/3205#issuecomment-482807885>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu5qNAx8jaNYWFXplUoFeRU6G9ZTeks5vgddRgaJpZM4cbpIE>
> .
>

huangruizhe commented:
Thanks! I have modified the cmd.sh and re-run the recipe for one day -- it looks like it will take much more time to finish all the iterations:
```
2019-04-13 21:25:49,434 [steps/nnet3/chain/e2e/train_e2e.py:381 - train - INFO ] Copying the properties from exp/chain/e2e_tdnnf_1a/egs to exp/chain/e2e_tdnnf_1a
2019-04-13 21:25:49,435 [steps/nnet3/chain/e2e/train_e2e.py:386 - train - INFO ] Preparing the initial acoustic model.
2019-04-13 21:25:53,674 [steps/nnet3/chain/e2e/train_e2e.py:414 - train - INFO ] Training will run for 10.0 epochs = 180 iterations
2019-04-13 21:25:53,674 [steps/nnet3/chain/e2e/train_e2e.py:460 - train - INFO ] Iter: 0/179   Jobs: 2   Epoch: 0.00/10.0 (0.0% complete)   lr: 0.001000
...
2019-04-14 10:31:48,726 [steps/nnet3/chain/e2e/train_e2e.py:460 - train - INFO ] Iter: 18/179   Jobs: 3   Epoch: 0.43/10.0 (4.3% complete)   lr: 0.001358
```

I think the modified parts in this PR (up to line 460) has been covered and there reports no error. So it may be ok to merge the PR (sorry for the delay). Otherwise, perhaps we will need to wait for the computation finishes (in 20 days estimated?).

kkm000 commented:
Wow! 20 days is a lot of entropy removed from the dataset into the environment in the form of heat! :)

I wound not really wait for completion. The diff from the initial change applied cleanly to all other files, and the added part (i. e. the `+++` of the diff) does not depend on any variables not in the `---` part; the code just copy-pasted initially. And you are right, the computation part has passed a few iterations.

kkm000 commented:
Out of curiosity, how many hours of data is crunched in this recipe at this stage (say, before the perturbace, just clean input)?

huangruizhe commented:
I think it should be the wsj dataset -- less than a hundred hours of speech, I suppose.
Btw, the computation becomes faster than last night. Now it has completed 33% of total iterations.
```
2019-04-14 21:54:27,182 [steps/nnet3/chain/e2e/train_e2e.py:460 - train - INFO ] Iter: 87/179   Jobs: 5   Epoch: 3.33/10.0 (33.3% complete)   lr: 0.001160
```
Perhaps this is due to the workload of our grid. It turns out that my estimation is not accurate.

kkm000 commented:
Is this ok to merge? No issues?

kkm000 commented:
Thank you @vimalmanohar for the review and @huangruizhe  for running the full test!
